### PR TITLE
Add factorized llama model for testing.

### PR DESCRIPTION
**Pull request by rjpower**
(Not intended for real review! But here's the current hacked up status of trying to get a layerwise trainer going in case it provides food for thought.) I can't for the life of me figure out how to get Github to acknowledge a file copy and show a diff: if there's a tip for how to do this with git, please let me know. Brain dump:

Random things:

* I noticed that `state.training_key` isn't updated with the `new_key` when we take a step: https://github.com/stanford-crfm/levanter/blob/main/src/levanter/trainer.py#L495 . IIUC, we'll use the same randomness for every batch as a result. (I could be horribly wrong here). That said, it doesn't look like we use dropout or any other runtime- randomness, so it probably doesn't change anything either way!

Good things:

* Setting up the factorized model was straightforward, especially thanks to the testing machinery: I could just iterate with pytest pretty quickly. The HF integration/state_dict was a bit harder (but definitely helped to have the roundtrip tests!). I banged my head on the layer linearization and trying to get the shapes aligned etc, but this pretty unusual setup though because the model weights radically diverge from the HF weights. In general I've found whacking around with the models a bit easier than mucking with the trainer setup.
* One nice thing that happened was by using the FactorizedLinear everywhere, I could centralize the HF encoding logic and avoid the flatten/unflatten logic in each transformer class. (I have a more grandiose thought of how we could factor out and maybe separately register the HF machinery but I think I should wait on that until I prove I actually know what I'm doing... 😛 )

There are a lot of my bad decisions here:

* I blundered a lot because I don't have much experience with JAX yet (!! yeah, I know)
* I made a copy of Trainer since I knew I was going to need hack around to understand things and get it working. Now that it's kind of working, I feel like you could *maybe* pull this off via subclassing, but it feels on the edge of understandability. I don't have any good ideas, but I think maybe we could find some components of trainer to make a bit more "top-level" and then re-use in different training paradigms as needed. I'll think on this some more: I'm thinking there's maybe a world where Trainer -> SupervisedModelTrainer but some of the ideas/components can be used elsewhere. This might be overkill too!
* There are dumb things like I pulled StepInfo and hooks, not for any good reason, but because I got a weird error around StepInfo and the JAX cache and I just wanted to get things working and then revisit.
* I had some confusion around getting the layerwise loss working correctly. It took me a bit to realize the default loss is hidden in the model and then I got tripped up by not setting the computation_axes correctly on my loss. This was all me: if I had bothered to read Trainer carefully I'd likely have understood the flow, but this part feels like something that would be easy to "lift" and share across different training paradigms.
* JAX/XLA was wickedly slow when I tried to compute loss & gradients for each layer incrementally, but at least for my small tests it did okay when I summed the individual layerwise losses: in theory I think this should give the same gradients/training behavior, but 🤷 . I disabled `scan_layers` to make it easy for me to invoke the layers individually, but from our discussion that cripples performance: I wonder if I need to be more clever here?

### CODE REVIEW

File: src/levanter/models/factorized_llama.py [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+        weights = state_dict[prefix + ".weight"]
+        down_proj, up_proj = low_rank_approximation(weights, self.down_proj.Out.size)
+        if self.out_first:
+            d[prefix + ".down_proj.weight"] = up_proj
+            d[prefix + ".up_proj.weight"] = down_proj
+        else:
+            d[prefix + ".down_proj.weight"] = down_proj
+            d[prefix + ".up_proj.weight"] = up_proj
+
+        d.update(
```
Comment (dlwh) [inline src/levanter/models/factorized_llama.py:232]: i need to finish the branch but this should be very simple once it's merged. You'll have to do your lowrank approximation manually but otherwise all of this will be handled for you.

In case you're curious/want to provide feedback, the branches are:

* https://github.com/stanford-crfm/haliax/pull/90
* https://github.com/stanford-crfm/levanter/pull/608/files

File: src/levanter/models/factorized_llama.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+    Vh_truncated = Vh[..., :rank, :]  # Note: Vh is already the conjugate transpose of V
+
+    # Reconstruct the low-rank approximation
+    down_proj = U_truncated @ S_truncated  # Shape (N, H)
+    up_proj = Vh_truncated  # Shape (H, M)
+
+    return down_proj, up_proj
+
+
+class FactorizedLinear(StateDictSerializationMixin, eqx.Module):
```
Comment (dlwh) [inline src/levanter/models/factorized_llama.py:193]: so one thing you could do is do all this as a tree transformation/model surgery, similar to how we do it for Lora.

Basically you load the model as normal, then you do something like:


```python

def replace_linear(x):
    if not isinstance(x, Linear): return x
    
    up_proj, down_proj = low_rank_approximation(x)
    return LowRankLinear(up_proj, down_proj)

modified_model = jax.tree_util.tree_map(replace_linear, model, is_leaf=lambda x: isinstance(x, Linear))
```

I'm pretty sure you could delete most of this file if you did this.

This version only works if use_scan_layers is false. You have to do some fanciness (that we do in Lora) for scan layers to essentially vmap the replace_linear layer whenever you detect a stacked. (Maybe I should make a stacked_aware_tree_map or something)

**Review by dlwh — COMMENTED** [review]
Sorry it took so long! Super busy week!

**Comment by dlwh** [comment]
> Random things:
> 
>     * I noticed that `state.training_key` isn't updated with the `new_key` when we take a step: https://github.com/stanford-crfm/levanter/blob/main/src/levanter/trainer.py#L495 . IIUC, we'll use the same randomness for every batch as a result. (I could be horribly wrong here). That said, it doesn't look like we use dropout or any other runtime- randomness, so it probably doesn't change anything either way!

Oops, that got lost in my last attempt to refactor. I should have a test for that.

Dropout is super slow on TPU so I tend to avoid it...

> 
> 
> Good things:
> 
>     * Setting up the factorized model was straightforward, especially thanks to the testing machinery: I could just iterate with pytest pretty quickly. The HF integration/state_dict was a bit harder (but definitely helped to have the roundtrip tests!). I banged my head on the layer linearization and trying to get the shapes aligned etc, but this pretty unusual setup though because the model weights radically diverge from the HF weights. In general I've found whacking around with the models a bit easier than mucking with the trainer setup.

Noted, the branches in Haliax/Levanter should make most of the "bookkeeping" parts go away, and I think if you moved to a "tree transformation" based approach, I think you shouldn't have to modify the core Llama code at all. (I think?!?)

> 
>     * One nice thing that happened was by using the FactorizedLinear everywhere, I could centralize the HF encoding logic and avoid the flatten/unflatten logic in each transformer class. (I have a more grandiose thought of how we could factor out and maybe separately register the HF machinery but I think I should wait on that until I prove I actually know what I'm doing... 😛 )

I'm all ears (though please do see the twin state dict branches and lmk what you think!)


> 
> 
> There are a lot of my bad decisions here:
> 
>     * I blundered a lot because I don't have much experience with JAX yet (!! yeah, I know)
> 
>     * I made a copy of Trainer since I knew I was going to need hack around to understand things and get it working. Now that it's kind of working, I feel like you could _maybe_ pull this off via subclassing, but it feels on the edge of understandability. I don't have any good ideas, but I think maybe we could find some components of trainer to make a bit more "top-level" and then re-use in different training paradigms as needed. I'll think on this some more: I'm thinking there's maybe a world where Trainer -> SupervisedModelTrainer but some of the ideas/components can be used elsewhere. This might be overkill too!
> 
>     * There are dumb things like I pulled StepInfo and hooks, not for any good reason, but because I got a weird error around StepInfo and the JAX cache and I just wanted to get things working and then revisit.
> 
>     * I had some confusion around getting the layerwise loss working correctly. It took me a bit to realize the default loss is hidden in the model and then I got tripped up by not setting the computation_axes correctly on my loss. This was all me: if I had bothered to read Trainer carefully I'd likely have understood the flow, but this part feels like something that would be easy to "lift" and share across different training paradigms.
> 
>     * JAX/XLA was wickedly slow when I tried to compute loss & gradients for each layer incrementally, but at least for my small tests it did okay when I summed the individual layerwise losses: in theory I think this should give the same gradients/training behavior, but 🤷 . I disabled `scan_layers` to make it easy for me to invoke the layers individually, but from our discussion that cripples performance: I wonder if I need to be more clever here?

**Comment by dlwh** [comment]
> There are a lot of my bad decisions here:
> 
>     * I blundered a lot because I don't have much experience with JAX yet (!! yeah, I know)

We all do! I think JAX needs better docs / examples for the "not a complete beginner but not a JAX whisperer" 

> 
>     * I made a copy of Trainer since I knew I was going to need hack around to understand things and get it working. Now that it's kind of working, I feel like you could _maybe_ pull this off via subclassing, but it feels on the edge of understandability. I don't have any good ideas, but I think maybe we could find some components of trainer to make a bit more "top-level" and then re-use in different training paradigms as needed. I'll think on this some more: I'm thinking there's maybe a world where Trainer -> SupervisedModelTrainer but some of the ideas/components can be used elsewhere. This might be overkill too!
> 

Yeah... still trying to figure this out. some bits like the batch loader helpers and some of the mixed precision/sharding logic should be "lego-ified" and split out. Probably other stuff too. I think my goal should be to make copy-paste not feel bad, if that makes sense.

>     * There are dumb things like I pulled StepInfo and hooks, not for any good reason, but because I got a weird error around StepInfo and the JAX cache and I just wanted to get things working and then revisit.

hrm. please send me a stack trace if you figure it out

> 
>     * I had some confusion around getting the layerwise loss working correctly. It took me a bit to realize the default loss is hidden in the model and then I got tripped up by not setting the computation_axes correctly on my loss. This was all me: if I had bothered to read Trainer carefully I'd likely have understood the flow, but this part feels like something that would be easy to "lift" and share across different training paradigms.
> 
>     * JAX/XLA was wickedly slow when I tried to compute loss & gradients for each layer incrementally, but at least for my small tests it did okay when I summed the individual layerwise losses: in theory I think this should give the same gradients/training behavior, but 🤷 . I disabled `scan_layers` to make it easy for me to invoke the layers individually, but from our discussion that cripples performance: I wonder if I need to be more clever here?

File: src/levanter/layerwise_trainer.py [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+}
+
+
+def _per_layer_loss(inputs, layer, key, mask):
+    loss, teacher_x = inputs
+    key, layer_k = jax.random.split(key, 2)
+    student, teacher = layer
+    student_y = student(teacher_x, mask=mask, key=layer_k)
+    teacher_y = teacher(teacher_x, mask=mask, key=layer_k)
+    loss = hax.mean((teacher_y - student_y) ** 2)
```
[the remark this answers is not in the document]
Comment (rjpower) [reply src/levanter/layerwise_trainer.py:112]: Oops, definitely -- lost that in the shuffle.

File: src/levanter/layerwise_trainer.py [hunk 4]
```diff
@@ -0,0 +1,740 @@
...
+        teacher: Teacher,
+        *args,
+        key: PRNGKeyArray,
+        is_trainable: FilterTree = True,
+        mp: Optional[jmp.Policy] = None,
+        fp8: Fp8Config = None,
+        **kwargs,
+    ) -> "TrainerState":
+        if mp is not None:
+            student = cast_params_by_trainability(student, mp, is_trainable)
```
[the remark this answers is not in the document]
Comment (rjpower) [reply src/levanter/layerwise_trainer.py:178]: Good idea, I didn't think too much about this one. We don't retain optimizer parameters for the teacher so I'm not too worried about space, but it's good for consistency.

File: src/levanter/layerwise_trainer.py [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+            key=key,
+            mask=batch.attn_mask,
+        )
+        return loss.scalar()
+
+
+class TrainerState(eqx.Module):
+    step: IntScalar = eqx.field(converter=_ensure_int_is_array)
+    student: Student
+    teacher: Teacher  # Can't be static, breaks tracing for some reason.
```
[the remark this answers is not in the document]
Comment (rjpower) [reply src/levanter/layerwise_trainer.py:148]: Ack, I sort of picked up on this over time. The meaning of _static_ is consistent, but subtle enough that I didn't quite follow it at first.

Interesting, having arrays be the signifier for dynamism would make sense too.

File: src/levanter/layerwise_trainer.py [hunk 6]
```diff
@@ -0,0 +1,740 @@
...
+        block_axis = student.transformer.layers.Block
+
+        def _is_block(leaf):
+            if hax.is_named_array(leaf):
+                print(leaf.shape, leaf.axes)
+            return hax.is_named_array(leaf) and leaf.axes[0] == block_axis
+
+        _loss_fn = hax.filter_checkpoint(_per_layer_loss)
+
+        loss, _ = hax.fold(
```
[the remark this answers is not in the document]
Comment (rjpower) [reply src/levanter/layerwise_trainer.py:132]: Yeah this took a bit of back and forth to figure out but seems sensible now. 

Initially I  used an unrolled loop, which "worked" on a tiny model but effectively stalled in compilation for the larger model sizes.
When I switched to scan, I ran into a few issues, most the normal JAX/TPU stuff, which shouldn't affect most regular models:

* The `is_scanned` magic for `hax.fold` was trying to scan over the `initial` values in addition to the layers (because it by default tries to glom any namedarray into the scan). Maybe it's worth exempting the first argument from that logic?
* It took a while to realize I had to enable gradient checkpointing. Everything "works" on a CPU test, but just OOMs the process. Trying it on a TPU yields XLA errors messages about trying to allocate >1TB of memory and asking for magic flags to actually tell you the large tensors. (And of course, when you provide those flags, it doesn't change anything...). I ended up needing to reduce the size of the model to convince XLA I was close enough to fitting for it to actually bother to do the allocation analysis and cough up the errors...

I'm also obviously special-casing to the `stacked` variant of the transformer, which feels a little gross, but... I suspect it's not worth generalizing the `hof.py` modules to support this.

File: src/levanter/models/factorized_llama.py (hunk 1 above)
Comment (rjpower) [reply src/levanter/models/factorized_llama.py:232]: Oh that's much nicer indeed. I like the separation of determining the saveable state and then transforming it to HF, and having `Stacked` et. al handle more of the load.

The linearization and `out_first` logic was definitely tedious to work with when I was trying to get this working. I ended up creating one of those weird models where every dimension is a prime number to figure out what was going where :). (It didn't help that I couldn't figure out the real shape that SVD was outputting...)

File: src/levanter/models/factorized_llama.py (hunk 2 above)
Comment (rjpower) [reply src/levanter/models/factorized_llama.py:193]: Oh that's a great idea. You'd lose a little flexibility in the size of your low-rank layers, but it would be nice to work across all models.

I'll try that out once I have something working end-to-end. (I don't mind starting with this and throwing it away, since it's a bit easier to debug this way than also debugging any wonky transform logic I try to write...

**Review by rjpower — COMMENTED** [review]
Thanks for the feedback! 

I think things are working, mostly, on the training side now: I can at least run steps. Unfortunately when I try to init the models from HF, things get stuck again for some reason. I'll need to add some more logging to see if I can figure out what's triggering that.

**Comment by rjpower** [comment]
> 
> Yeah... still trying to figure this out. some bits like the batch loader helpers and some of the mixed precision/sharding logic should be "lego-ified" and split out. Probably other stuff too. I think my goal should be to make copy-paste not feel bad, if that makes sense.

+1,  and overall I think things have been fine: if it's been slow getting things working, it's because I hit some indecipherable XLA or JAX error vs having to hack around with Levanter. If this was just another model I don't see any reason I'd presumably be able to reuse the default training setup easily. But I suspect there are some parts here we can make easier to use while preserving the existing workflow.

> 
> hrm. please send me a stack trace if you figure it out

I'll let you know if I can reproduce. I couldn't understand it because all that stuff is outside of the JIT scope, but I thought, "no need for StepInfo until I can take a step..."


> > * JAX/XLA was wickedly slow when I tried to compute loss & gradients for each layer incrementally, but at least for my small tests it did okay when I summed the individual layerwise losses: in theory I think this should give the same gradients/training behavior, but 🤷 . I disabled `scan_layers` to make it easy for me to invoke the layers individually, but from our discussion that cripples performance: I wonder if I need to be more clever here?
> > ```
> 
> hrm. I do think the timings in the compiler are pretty carefully tuned for the "standard" case and maybe you were tripping it up? I do wish they'd offered PGO for TPU like they're starting to do for GPU... though maybe that's not the problem here.

Yeah there's certainly some tuning issues with the compiler if you don't take the right approach. I didn't investigate it too much in this case though: it might just have produced a giant program and stalled trying to compile it...

**Comment by dlwh** [comment]
> The is_scanned magic for hax.fold was trying to scan over the initial values in addition to the layers (because it by default tries to glom any namedarray into the scan). Maybe it's worth exempting the first argument from that logic?

Can you open an issue. I agree this is better
